File: 63-unique-paths-ii/unique-paths-ii.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int [][] dp = new int[obstacleGrid.length+1][obstacleGrid[0].length+1];
        for(int [] arr : dp){
            Arrays.fill(arr,-1);
        }
        return fun(0,0,obstacleGrid,dp);
    }

    public int fun(int i,int j, int [][] obstacleGrid,int [][] dp){
        if(i >= obstacleGrid.length || j >= obstacleGrid[0].length || obstacleGrid[i][j] == 1){
            return 0;
        }
        if(i == obstacleGrid.length-1 && j == obstacleGrid[0].length-1){
            return 1;
        }
        if(dp[i][j] != -1){
            return dp[i][j];
        }
        int right = fun(i+1,j,obstacleGrid,dp);
        int down = fun(i,j+1,obstacleGrid,dp);
        dp[i][j] = right + down;
        return dp[i][j];
    }
}

File: 63-unique-paths-ii/unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        n = len(obstacleGrid)
        m = len(obstacleGrid[0])
        if obstacleGrid[0][0] == 1 or obstacleGrid[n-1][m-1] == 1:
            return 0
        dirs = [(1,0),(0,1)]
        dp = [[-1 for _ in range(m+1)] for _ in range(n+1)]
        def fun(i,j):
            if i == n-1 and j == m-1:
                return 1
            if dp[i][j] != -1:
                return dp[i][j]
            ans = 0
            for di,dj in dirs:
                ni,nj = i + di, j + dj
                if 0 <= ni < n and 0 <= nj < m and obstacleGrid[ni][nj] != 1:
                    ans += fun(ni,nj)
            dp[i][j] = ans
            return ans
        return fun(0,0)
